packages/core/src/types.ts

```diff
@@ -75,6 +73,381 @@ export interface PositionWithOffset extends Position {
+	 * @param syntaxElement A syntax element with location information to replace.
+	 * @param text The text to replace the node or token with.
+	 * @returns The edit to replace the node or token.
+	 */
+	replaceText(
+		syntaxElement: EditableSyntaxElement,
+		text: string,
+	): RuleTextEdit;
+
+	/**
+	 * Replaces the specified range with the given text.
+	 * @param range The range to replace.
+	 * @param text The text to replace the range with.
+	 * @returns The edit to replace the range.
+	 */
+	replaceTextRange(range: SourceRange, text: string): RuleTextEdit;
+}
+
+/**
+ * Represents a fix for a rule violation implemented as a text edit.
+ */
+export interface RuleTextEdit {
+	/**
+	 * The range to replace.
+	 */
+	range: SourceRange;
+
+	/**
+	 * The text to insert.
+	 */
+	text: string;
+}
+
+// #endregion
+
+interface ViolationReportBase {
+	/**
+	 * The type of node that the violation is for.
+	 * @deprecated May be removed in the future.
+	 */
+	nodeType?: string | undefined;
+
+	/**
+	 * The data to insert into the message.
+	 */
+	data?: Record<string, string> | undefined;
+
+	/**
+	 * The fix to be applied for the violation.
+	 * @param fixer The text editor to apply the fix.
+	 * @returns The fix(es) for the violation.
+	 */
+	fix?(
+		fixer: RuleTextEditor,
+	): RuleTextEdit | RuleTextEdit[] | IterableIterator<RuleTextEdit> | null;
+
+	/**
+	 * An array of suggested fixes for the problem. These fixes may change the
+	 * behavior of the code, so they are not applied automatically.
+	 */
+	suggest?: SuggestedEdit[];
+}
+
+type ViolationMessage = { message: string } | { messageId: string };
+type ViolationLocation<Node> = { loc: SourceLocation } | { node: Node };
+
+export type ViolationReport<Node = unknown> = ViolationReportBase &
+	ViolationMessage &
+	ViolationLocation<Node>;
+
+// #region Suggestions
+
+interface SuggestedEditBase {
+	/**
+	 * The data to insert into the message.
+	 */
+	data?: Record<string, string> | undefined;
+
+	/**
+	 * The fix to be applied for the suggestion.
+	 * @param fixer The text editor to apply the fix.
+	 * @returns The fix for the suggestion.
+	 */
+	fix?(
+		fixer: RuleTextEditor,
+	): RuleTextEdit | RuleTextEdit[] | IterableIterator<RuleTextEdit> | null;
+}
+
+type SuggestionMessage = { desc: string } | { messageId: string };
+
+/**
+ * A suggested edit for a rule violation.
+ */
+export type SuggestedEdit = SuggestedEditBase & SuggestionMessage;
+
+// #endregion
+
+/**
+ * The definition of an ESLint rule.
+ */
+export interface RuleDefinition<
+	LangOptions = LanguageOptions,
+	Code extends SourceCode = SourceCode,
+	RuleOptions = unknown[],
+	Visitor extends RuleVisitor = RuleVisitor,
+> {
+	/**
+	 * The meta information for the rule.
+	 */
+	meta?: RulesMeta;
+
+	/**
+	 * Creates the visitor that ESLint uses to apply the rule during traversal.
+	 * @param context The rule context.
+	 * @returns The rule visitor.
+	 */
+	create(context: RuleContext<LangOptions, Code, RuleOptions>): Visitor;
```

Comment:
> `context: RuleContext<...>`

This isn't passing an argument for the `Node = unknown` type parameter. Is it intentional that `RuleDefinition` wouldn't know about the type of nodes in its language?

Reply:
Not intentional, just missed it. Thanks!